proposals/per-extension-language-preferences.md

```diff
@@ -0,0 +1,234 @@
+# Proposal: Per-extension language preferences
+
+**Summary**
+
+This proposal allows developers (users) to set a specific language for their extension (which may be different from the browser or system default language) and create a language selection menu for users.
+
+**Document Metadata**
+
+**Author:** hanguokai
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** 
+
+**Created:** 2024-06-16
+
+**Related Issues:** [#258](https://github.com/w3c/webextensions/issues/258)
+
+## Motivation
+
+### Objective
+
+Prior to this proposal, browsers automatically selected a `browser.i18n` language for extensions, and developers (users) could not set a specific language.
+
+For multilingual users, sometimes they want to set a specific application (extension) to another language that is different from the system (browser). This need is not uncommon.
+[Android 13](https://developer.android.com/guide/topics/resources/app-languages), [iOS 13 and macOS Catalina](https://developer.apple.com/news/?id=u2cfuj88) support a feature
+called "Per-app language preferences (settings)". This proposal brings the same function to browser extensions.
+
+#### Use Cases
+
+- Extensions create a language selection menu for users.
+- [Optional] Browsers could potentially provide a built-in language selection menu for users, just like Android and iOS.
+- [Optional] Extensions could integrate third-party i18n implementations (frameworks) with the browser's built-in language selection menu.
+
+### Known Consumers
+
+This is a generic i18n feature, which means that all types of extensions can use it. Some extensions already provide this functionality via workarounds or non-browser.i18n implementations.
+
+## Specification
+
+### Definitions
+
+In this document:
+- **locale** or **language code** or **language tag** is a string that represent a language, defined in [BCP 47](https://www.rfc-editor.org/info/bcp47).
+For example, `en-US`, `zh-CN`, `fr`. It is used by `browser.i18n`, `Date`, `Intl` and various other APIs.
```

Comment:
`zh-CN` is presumably meant to indicate Simplified Chinese, which is also used in Singapore. It might be better to use `zh-Hans` as the language tag in the example.

See also https://www.w3.org/International/articles/language-tags/index.en.html#script

Reply:
Perhaps due to historical reasons, browsers still uses 'zh-CN' instead of 'zh-Hans'. For example, `navigator.language` returns "zh-CN" in Chrome and Firefox on macOS. And Chrome only supports [a list of languages](https://developer.chrome.com/docs/extensions/reference/api/i18n#locales) in the `/_locates/` directory, which only contains a limited combination of languages and regions.

I think solving this problem seems to go beyond the core issue that this proposal aims to address. This proposal is intended to expand functionality on the basis of existing capabilities.

Reply:
Still would be good and is also relevant for this issue what locales a browser supports / can deal with. See also: https://github.com/w3c/webextensions/issues/131

For example, the Naver Whale store rejects any language tag which includes the script subtag (Hans, Latn, Cyrl, and others).